### PR TITLE
Update opentelemetry to version 0.21 (+ stuff that was in flight)

### PR DESCRIPTION
**Pull request by garypen**
Updates the other otel ecosystem crates as well.

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [x] Changes are compatible[^1]
- [x] Documentation[^2] completed
- [x] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [x] Unit Tests
    - [x] Integration Tests
    - [x] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

### CODE REVIEW

File: apollo-router/src/plugins/telemetry/formatters/json.rs [hunk 1]
```diff
@@ -124,39 +124,41 @@ where
...
-                    &opentelemetry::Value,
-                > = custom_attributes.iter().collect();
-                for (key, value) in custom_attributes.iter().filter(|(k, _)| {
-                    let key_name = k.as_str();
+                let custom_attributes: Vec<&opentelemetry::KeyValue> = {
+                    let mut my_custom_attributes: Vec<&opentelemetry::KeyValue> =
+                        custom_attributes.iter().collect();
+                    my_custom_attributes.sort_by_key(|kv| &kv.key);
+                    my_custom_attributes
+                };
```
Comment (bnjjj) [inline apollo-router/src/plugins/telemetry/formatters/json.rs:132]: I think this part is useless now as custom_attributes are already a vec right ?

File: apollo-router/src/plugins/telemetry/formatters/text.rs [hunk 2]
```diff
@@ -218,8 +218,14 @@ impl Text {
                 write!(writer, "{}{{", span.name())?;
             }
             #[cfg(test)]
-            let attrs: BTreeMap<&opentelemetry::Key, &opentelemetry::Value> = attrs.collect();
-            for (key, value) in attrs {
+            let attrs: Vec<&opentelemetry::KeyValue> = {
+                let mut my_attrs: Vec<&opentelemetry::KeyValue> = attrs.collect();
+                my_attrs.sort_by_key(|kv| &kv.key);
+                my_attrs
+            };
```
Comment (bnjjj) [inline apollo-router/src/plugins/telemetry/formatters/text.rs:225]: Same here

File: apollo-router/src/plugins/telemetry/formatters/text.rs [hunk 3]
```diff
@@ -241,8 +247,14 @@ impl Text {
...
```
Comment (bnjjj) [inline apollo-router/src/plugins/telemetry/formatters/text.rs:254]: same

**Review by bnjjj — COMMENTED** [review]
Perfs are better now, once you deleted the useless parts we're good :)

File: apollo-router/src/plugins/telemetry/formatters/json.rs (hunk 1 above)
Comment (garypen) [reply apollo-router/src/plugins/telemetry/formatters/json.rs:132]: It's required to sort or the insta tests won't pass. This code is all #[cfg(test)]

File: apollo-router/src/plugins/telemetry/formatters/text.rs (hunk 2 above)
Comment (garypen) [reply apollo-router/src/plugins/telemetry/formatters/text.rs:225]: as above.

File: apollo-router/src/plugins/telemetry/formatters/text.rs (hunk 3 above)
Comment (garypen) [reply apollo-router/src/plugins/telemetry/formatters/text.rs:254]: as above.

**Review by bnjjj — APPROVED** [review]
Ignore my comments, I'll clean it later

File: apollo-router/src/plugins/telemetry/formatters/json.rs (hunk 1 above)
Comment (Geal) [reply apollo-router/src/plugins/telemetry/formatters/json.rs:132]: good idea to mark it cfg(test), it won't have any impact otherwise

**Review by Geal — APPROVED** [review]

**Pull request merged**